src/rust/engine/src/rule_graph.rs

```diff
@@ -159,6 +160,45 @@ impl <'a> GraphMaker<'a> {
     GraphMaker { tasks: tasks, root_subject_types: root_subject_types }
   }
 
+  pub fn sub_graph(&self, subject_type: &TypeId, product_type: &TypeConstraint) -> RuleGraph {
+    let mut full_root_rule_dependency_edges: RootRuleDependencyEdges = HashMap::new();
+    let mut full_dependency_edges: RuleDependencyEdges = HashMap::new();
+    let mut full_unfulfillable_rules: UnfulfillableRuleMap = HashMap::new();
+
+    let beginning_root_opt = self.gen_root_entry(subject_type, product_type);
+    if beginning_root_opt.is_none() {
+      return RuleGraph {   root_subject_types: vec![],
+        root_dependencies: full_root_rule_dependency_edges,
+        rule_dependency_edges: full_dependency_edges,
+        unfulfillable_rules: full_unfulfillable_rules,
+      }
+    }
+    let beginning_root = beginning_root_opt.unwrap();
+
+    let constructed_graph = self._construct_graph(
+      beginning_root,
+      full_root_rule_dependency_edges,
+      full_dependency_edges,
+      full_unfulfillable_rules
+    );
+
+    // less than ideal, the copying
+    full_root_rule_dependency_edges = constructed_graph.root_dependencies.clone();
+    full_dependency_edges = constructed_graph.rule_dependency_edges.clone();
+    full_unfulfillable_rules = constructed_graph.unfulfillable_rules.clone();
+
+    self.add_unreachable_rule_diagnostics(&full_dependency_edges, &mut full_unfulfillable_rules);
```

Comment:
It certainly seems like you should be able to pass these in without cloning them... is the issue just that `constructed_graph` isn't mutable? `let mut constructed_graph = ...`